### DISCUSSION TITLE
Looking for a lazy (delayed) loading solution.

### DISCUSSION THREAD

**Discussion by jcejohnson** [pydantic/pydantic#5140] [Question] [answered]
Let's say I have a ginormous json document that is saved as dozens or hundreds of small(ish) json files linked together using jsonref. Is there a way to tell pydantic to only load and validate the files as they are accessed?

For instance, I have a (contrived) ActorsDatabase model with a collection of actors, their movies, etc. saved such that each movie an actor appears in is a separate json file. What I would like to see happen is for the file data/actors/charlie_chaplin/movies/modern_times.json to be lazy loaded and validated when the code does `modern_times = the_model.actors["charlie_chaplin"].movies["modern_times"]`

**Comment by samuelcolvin** [comment, marked as answer]
Should be easy enough to set this up, either with a simple function

`get_move(query)`

Which gets the data from a dict, file or DB, then runs validation.

**Reply by jcejohnson** [reply]
If I'm following you then that would put the lazy-loading into the code that uses the model. What I'm hoping to do is push the lazy-loading into the model itself so that the lazy bit is transparent to the code using the model.

That is:
```
# Load & validate "actor_data/root.json"  ->  { "actors": { "$ref": "root/actors.json } }
the_model = ActorsDatabase.parse_file("actor_data/root.json")

# Load & validate each of:
# - "actor_data/root/actors.json"  ->    { "charlie_chaplin": { "$ref": "actors/charlie_chaplin.json" }, ... }
# - "actor_data/root/actors/charlie_chaplin.json"  ->  { "movies": { "$ref": "charlie_chaplin/movies.json" }, "fn": "Charlie", ... }
# - "actor_data/root/actors/charlie_chaplin/movies.json"  ->  ...
# - "actor_data/root/actors/charlie_chaplin/movies/modern_times.json"  ->  ...
modern_times = the_model.actors["charlie_chaplin"].movies["modern_times"]
```
I guess I'm asking if there is a way to delay `...validate_python(data)` in `BaseModel.__init__()` until any of the attributes are referenced but that sounds like a bad idea.

I suppose that instead of defining the model as:
```
class ActorsDatabase(BaseModel):
    actors: Dict[str, Actor]
```
I could have separate models for an ActorReference and Actor and Union them in the Dict.
e.g.:
```
class ActorRef(BaseModel):
    ref: str = Field(alias="$ref")
    _target: Actor = None
    def __getattr__(self, name):
        self._target = self._target or Actor.parse_raw(self._load_reference())
        return self._target.__getattr__(name)
    def self._load_reference(self):
        # Magic goes here
    
class ActorsDatabase(BaseModel):
    actors: Dict[str, Union[ActorRef, Actor]]
```

That would make the model a little messy but that's probably OK.

**Reply by samuelcolvin** [reply]
Well if your really wanted you could implement your own type that implements `__getitem__` and does the validation then, but that doesn't require anything special in pydantic - the pydantic model is just an attribute of your special magic dict-like thing, which you use in `__getitem__`.

**Reply by jcejohnson** [reply]
I like that idea! In fact, I've already got some custom `__getitem__` stuff going on that I may be able to enhance.
